### PR TITLE
Martin/ios 491 double tap to expand input

### PR DESCRIPTION
**Pull request by MartinJibber**
https://user-images.githubusercontent.com/88675954/155436849-b35ee12f-cba6-4b60-a703-eafb4e27c87d.mp4

### CODE REVIEW

File: Shared/Extensions/Selectable.swift [hunk 1]
```diff
@@ -146,28 +146,60 @@ class PanGestureRecognizer: UIPanGestureRecognizer {
     }
 }
 
+/// Similar to pan gesture recognizer except that the begin state is triggered after a short delay, even if the touch doesn't move.
```
Comment (MartinJibber) [inline Shared/Extensions/Selectable.swift:149]: Now we don't immediately start the pan gesture for swiping up a message. This helps reduce conflicts with double tap gestures.

File: Shared/UI/Views/TextView.swift [hunk 2]
```diff
@@ -14,16 +14,18 @@ class TextView: UITextView {
 
     // Trim white space and newline characters when end editing. Default is true
     var trimWhiteSpaceWhenEndEditing: Bool = true
-
     // Maximum length of text. 0 means no limit.
-    var maxLength: Int = 250
+    var maxLength: Int = 500
```
Comment (MartinJibber) [inline Shared/UI/Views/TextView.swift:18]: Increasing the character limit.

File: Shared/UI/Views/TextView.swift [hunk 3]
```diff
@@ -14,16 +14,18 @@ class TextView: UITextView {
...
+
+    @Published var isEditing: Bool = false
```
Comment (MartinJibber) [inline Shared/UI/Views/TextView.swift:20]: Allowing outside classes to know when the text view starts editing.

File: Shared/UI/Views/TextView.swift [hunk 4]
```diff
@@ -14,16 +14,18 @@ class TextView: UITextView {
...
 
-    var cancellables = Set<AnyCancellable>()
+    private var cancellables = Set<AnyCancellable>()
 
     var numberOfLines: Int {
         guard let lineHeight = self.font?.lineHeight else { return 0 }
-        let height = self.contentSize.height - self.textContainerInset.top - self.textContainerInset.bottom
-        return Int(height / lineHeight)
```
Comment (MartinJibber) [inline Shared/UI/Views/TextView.swift:26]: For whatever reason this line count logic was unreliable. The new logic seems to work really well.

File: Shared/UI/Views/TextView.swift [hunk 5]
```diff
@@ -137,7 +139,25 @@ class TextView: UITextView {
             }.store(in: &self.cancellables)
     }
 
-    // MARK: Setters
+    override func draw(_ rect: CGRect) {
```
Comment (MartinJibber) [inline Shared/UI/Views/TextView.swift:142]: Reorganization

File: Shared/UI/Views/TextView.swift [hunk 6]
```diff
@@ -190,71 +210,55 @@ class TextView: UITextView {
         self.textDidChange()
     }
 
-    // Trim white space and new line characters when end editing.
+    // MARK: - TextView Event Handlers
+
+    func textViewDidBeginEditing() {
+        self.isEditing = true
+    }
```
Comment (MartinJibber) [inline Shared/UI/Views/TextView.swift:217]: Let subscribers know about editing state changes.

File: Shared/UI/Views/TextView.swift [hunk 7]
```diff
@@ -265,4 +269,16 @@ class TextView: UITextView {
...
+        guard let text = self.text, !text.isEmpty, let attText = self.attributedText else {
+            return CGSize.zero
+        }
+
+        let maxTextSize = CGSize(width: maxWidth, height: maxHeight)
+
+        return attText.boundingRect(with: maxTextSize,
+                                    options: .usesLineFragmentOrigin,
+                                    context: nil).size
+    }
```
Comment (MartinJibber) [inline Shared/UI/Views/TextView.swift:283]: Separated out the text size logic so I could use it in the line count getter.

File: iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift [hunk 8]
```diff
@@ -36,8 +36,8 @@ class ConversationListViewController: ViewController, ConversationListCollection
     var onSelectedMessage: ((_ cid: ChannelId, _ messageId: MessageId, _ replyId: MessageId?) -> Void)?
 
     // Custom Input Accessory View
-    lazy var messageInputAccessoryView: ConversationInputAccessoryView = {
-        let inputView: ConversationInputAccessoryView = ConversationInputAccessoryView.fromNib()
+    lazy var messageInputAccessoryView: SwipeableInputAccessoryView = {
+        let inputView: SwipeableInputAccessoryView = SwipeableInputAccessoryView.fromNib()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation List/ConversationListViewController.swift:40]: No need to have a subclass any more.

File: iOS/Flows/Conversation/Input/InputTextView.swift [hunk 9]
```diff
@@ -15,7 +15,6 @@ enum InputType {
     case keyboard
     case calendar
     case jibs
-    case confirmation
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Input/InputTextView.swift:18]: Not used.

File: iOS/Flows/Conversation/Input/PreviewMessageView.swift [hunk 10]
```diff
@@ -60,7 +61,7 @@ class PreviewMessageView: SpeechBubbleView {
         self.imageView.height = self.imageView.displayable.isNil ? 0 : 100
         self.imageView.centerOnX()
         
-        self.textView.setSize(withMaxWidth: self.width)
+        self.textView.setSize(withMaxWidth: self.width, maxHeight: self.height)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Input/PreviewMessageView.swift:64]: Fixing how the preview message view handles long text blocks.

File: iOS/UI/Swipeable Input/SwipeInputHintAnimator.swift [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+//  Jibber
+//
+//  Created by Martin Young on 2/23/22.
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+
+/// Manages hint animations on behalf of a SwipeableInputAccessoryView.
+class SwipeInputHintAnimator {
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeInputHintAnimator.swift:12]: Moved the hint animation logic into a support class.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.xib [hunk 12]
```diff
@@ -16,7 +16,7 @@
     <objects>
         <placeholder placeholderIdentifier="IBFilesOwner" id="-1" userLabel="File's Owner"/>
         <placeholder placeholderIdentifier="IBFirstResponder" id="-2" customClass="UIResponder"/>
-        <view contentMode="scaleToFill" id="iN0-l3-epB" customClass="ConversationInputAccessoryView" customModule="Jibber" customModuleProvider="target">
+        <view contentMode="scaleToFill" id="iN0-l3-epB" customClass="SwipeableInputAccessoryView" customModule="Jibber" customModuleProvider="target">
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.xib:19]: Made a bunch of constraints adjustments to get this working.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 13]
```diff
@@ -31,23 +30,33 @@ protocol SwipeableInputAccessoryViewDelegate: AnyObject {
...
+class SwipeableInputAccessoryView: BaseView {
 
     weak var delegate: SwipeableInputAccessoryViewDelegate?
 
+    enum InputState {
+        /// The input field is fit to the current input. Swipe to send is enabled.
+        case collapsed
+        /// The input field is expanded and can be tapped to edit/copy/paste. Swipe to send is disabled.
+        case expanded
+    }
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:42]: Swipe Accessory now has two states.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 14]
```diff
@@ -76,7 +88,16 @@ class SwipeableInputAccessoryView: BaseView, UIGestureRecognizerDelegate, Active
...
 
-    var cancellables = Set<AnyCancellable>()
+    private var cancellables = Set<AnyCancellable>()
+
+    // MARK: - Layout/Animation Properties
+
+    @IBOutlet var textViewCollapsedVerticalCenterConstraint: NSLayoutConstraint!
+    @IBOutlet var textViewCollapsedMaxHeightConstraint: NSLayoutConstraint!
+    @IBOutlet var textViewExpandedTopPinConstraint: NSLayoutConstraint!
+    @IBOutlet var textViewExpandedBottomPinConstraint: NSLayoutConstraint!
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:98]: The text view needs different constraints depending on the state we're in.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 15]
```diff
@@ -126,64 +147,148 @@ class SwipeableInputAccessoryView: BaseView, UIGestureRecognizerDelegate, Active
...
+    private func updateInputState(with numberOfLines: Int) {
+        // When the text hits 3 lines, transition to the expanded state.
+        // However don't automatically go back to the collapsed state when the line count is less than 3.
+        guard numberOfLines > 2 else { return }
+        self.inputState = .expanded
+    }
+
+    private func updateLayout(for inputState: InputState) {
+        let newInputHeight: CGFloat
+
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:251]: Whenever the input state changes, this configuration logic gets triggered.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 16]
```diff
@@ -232,126 +318,20 @@ class SwipeableInputAccessoryView: BaseView, UIGestureRecognizerDelegate, Active
...
     }
 
     func updateInputType(with type: InputType) {
         self.textView.updateInputView(type: type)
     }
 
-    func animateInputViews(with text: String) {}
-
     func resetInputViews() {
+        self.inputState = .collapsed
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:336]: Return back to the collapsed state.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 17]
```diff
@@ -232,126 +318,20 @@ class SwipeableInputAccessoryView: BaseView, UIGestureRecognizerDelegate, Active
...
-
-    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer,
-                           shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer) -> Bool {
-
-        if gestureRecognizer is UIPanGestureRecognizer {
-            return false
-        }
-
-        return true
-    }
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:354]: No longer needed.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 18]
```diff
@@ -126,64 +147,148 @@ class SwipeableInputAccessoryView: BaseView, UIGestureRecognizerDelegate, Active
...
+            // If we are editing, a double tap should trigger the expanded state.
+            // If we're not editing, it takes 1 tap to start.
+            self.inputFieldTapRecognizer.numberOfTapsRequired = isEditing ? 2 : 1
+        }.store(in: &self.cancellables)
+
+        self.$inputState
+            .removeDuplicates()
+            .mainSink { [unowned self] inputState in
+                self.updateLayout(for: inputState)
+            }.store(in: &self.cancellables)
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:237]: Publishers are nice. Whenever the input state changes, we guarantee the update layout function is called.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 19]
```diff
@@ -126,64 +147,148 @@ class SwipeableInputAccessoryView: BaseView, UIGestureRecognizerDelegate, Active
...
-        if numberOfLines > 3 {
-            new = self.textView.height + self.inputTypeContainer.height
+
+    // MARK: - State Updates
+
+    private func updateInputState(with numberOfLines: Int) {
+        // When the text hits 3 lines, transition to the expanded state.
+        // However don't automatically go back to the collapsed state when the line count is less than 3.
+        guard numberOfLines > 2 else { return }
+        self.inputState = .expanded
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:246]: See comment.

File: iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift [hunk 20]
```diff
@@ -126,64 +147,148 @@ class SwipeableInputAccessoryView: BaseView, UIGestureRecognizerDelegate, Active
...
+
+    private func handleInputTap() {
+        if self.textView.isFirstResponder {
+            // When the text view is editing, double taps should expand it.
+            self.inputState = .expanded
+        } else {
+            // If we're not editing, a tap starts editing.
+            self.textView.updateInputView(type: .keyboard, becomeFirstResponder: true)
+        }
+    }
```
Comment (MartinJibber) [inline iOS/UI/Swipeable Input/SwipeableInputAccessoryView.swift:190]: Tapping the input now has two different behaviors.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**